src/Expander.hs

```diff
@@ -688,22 +688,24 @@ initializeKernel outputChannel = do
                  , _datatypeName = dn
                  }
       let tyImpl =
-             \k dest stx -> do
-               equateKinds stx KStar k
-               Stx _ _ (me, args) <- mustBeCons stx
-               _ <- mustBeIdent me
-               if length args /= length argKinds
-                 then throwError $ WrongDatatypeArity stx dt
-                                     (fromIntegral $ length argKinds)
-                                     (length args)
-                 else do
-                   argDests <- traverse (uncurry scheduleType) (zip argKinds args)
-                   linkType dest $ tDatatype dt argDests
+            \k dest stx -> do
+              Stx _ _ (me, args) <- mustBeCons stx
+              _ <- mustBeIdent me
+              if length args > length argKinds
```

Comment:
Now accepting the case in which fewer arguments are provided than required, e.g. `(Either (Unit))` instead of `(Either (Unit) (Unit))`.

Reply:
Why is it that these types explicitly need to check for too many arguments? Wouldn't the kind system figure that part out? (Basically, isn't `WrongDatatypeArity` a thing made obsolete by the introduction of kinds other than `*`?)

Reply:
Oh nice, I didn't realize that check was no longer necessary!

Reply:
Well, it's only no longer necessary if there's not a missing check in the kind checker :-) It's worth testing that case if it's not already there.

Reply:
Alas, without this `if length args > length argKinds` check, the following is accepted. I have just created #194 to track this.

```
-- (just unit) : (Maybe Unit) 
(example
  (the (Maybe Unit Unit)
       (just unit)))
```

Reply:
What about the similar check a few lines below, for type patterns? A value of type `Type` could hold a type of a kind other than `*` (as demonstrated by the variable named `type` in this PR's `higher-kinded-patterns.kl` example, which holds a type of kind `* -> *`), so the kind checker doesn't have much information with which to perform its check.

Reply:
It seems plausible to me that type patterns should be well-kinded, and that this should be caught. What I'm less sure of is whether they should all have kind *, or whether it makes sense to match against other types during expansion.

Reply:
I have created #196 to continue the discussion on type patterns after I merge this PR.